File: Regine/regine.py
#solve 8 queens problem using A* algorithm

# return the number of pairs of queens that are attacking each other
def heuristic(state):
    pairs = 0
    for i in range(len(state)):
        for j in range(i+1, len(state)):
            if state[i] == state[j]:
                pairs += 1
            if abs(state[i] - state[j]) == abs(i - j):
                pairs += 1
    return pairs

def AStarSearch(state, heuristic):
    open = [state]
    closed = []
    while open:
        open.sort(key=heuristic)
        state = open.pop(0)
        closed.append(state)
        if heuristic(state) == 0:
            return state
        for i in range(len(state)):
            for j in range(len(state)):
                if state[i] != j: # if the queen is not already in this column
                    newState = state[:] # copy the state
                    newState[i] = j
                    print(newState)
                    if newState not in open and newState not in closed:
                        open.append(newState)
    return None

def main():
    state = [0,1,2,3]
    solution = AStarSearch(state, heuristic)
    if solution:
        print(solution)
    else:
        print("No solution found")

if __name__ == "__main__":
    main()

File: Regine/regine.pl
:- use_module(library(lists)).

% Heuristic function that returns the number of queen pairs that threaten each other
/*heuristic(State, H) :-
    findall((I, J), (nth0(I, State, X), nth0(J, State, Y), I \= J, abs(X - Y) =:= abs(I - J)), Threats),
    length(Threats, H1),
    length(State, N),
    H is H1 + N.
*/

% A* search algorithm
a_star(Start, Path) :-
    heuristic(Start, H),
    a_star([(Start, H, 0)], [], Path).

a_star([(State, _, G) | _], _, Path) :-
    goal(State),
    reverse(G, Path).

a_star([(State, H, G) | Open], Closed, Path) :-
    findall((Next, H1, G1), (succ(State, Next), \+ member((Next, _, _), Closed),
    G1 is G + 1, heuristic(Next, H1)), Successors),
    append(Successors, Open, Open1),
    sort(Open1, Open2),
    a_star(Open2, [(State, H, G) | Closed], Path).

% Successor function that generates all valid next states
succ(State, Next) :-
    length(State, N),
    nth1(N, State, X),
    between(1, N, Y),
    \+ member(Y, State),
    X1 is X + 1,
    Y1 is Y - 1,
    Y2 is Y + 1,
    (X1 =:= Y ; X1 =:= Y1 ; X1 =:= Y2),
    replace(State, N, Y, Next).

% Replace the N-th element of a list with a new value
replace([_|T], 1, X, [X|T]).
replace([H|T], N, X, [H|R]) :-
    N > 1,
    N1 is N - 1,
    replace(T, N1, X, R).

% Goal predicate that checks if all queens are placed without threatening each other
goal(State) :-
    minacciate(R, State),
    length(R, 0).

% Test the algorithm with a 4x4 board
test() :-
    a_star([2, 4, 1, 3], Path),
    writeln(Path).